Mage.Tests/src/test/java/org/mage/test/player/RandomPlayer.java

```diff
@@ -75,42 +75,25 @@ public boolean priority(Game game) {
     private Ability getAction(Game game) {
         List<ActivatedAbility> playables = getPlayableAbilities(game);
         Ability ability;
-        while (true) {
-            if (playables.size() == 1) {
-                ability = playables.get(0);
+        if (playables.size() == 1) {
+            ability = playables.get(0);
+        } else {
+            ability = playables.get(RandomUtil.nextInt(playables.size()));
+        }
+        List<Ability> options = getPlayableOptions(ability, game);
+        if (!options.isEmpty()) {
+            if (options.size() == 1) {
+                ability = options.get(0);
             } else {
-                ability = playables.get(RandomUtil.nextInt(playables.size()));
+                ability = options.get(RandomUtil.nextInt(options.size()));
             }
-            List<Ability> options = getPlayableOptions(ability, game);
-            if (!options.isEmpty()) {
-                if (options.size() == 1) {
-                    ability = options.get(0);
-                } else {
-                    ability = options.get(RandomUtil.nextInt(options.size()));
-                }
-            }
-            if (!ability.getManaCosts().getVariableCosts().isEmpty()) {
-                int amount = getAvailableManaProducers(game).size() - ability.getManaCosts().manaValue();
-                if (amount > 0) {
-                    ability = ability.copy();
-                    ability.getManaCostsToPay().add(new GenericManaCost(RandomUtil.nextInt(amount)));
-                }
+        }
+        if (!ability.getManaCosts().getVariableCosts().isEmpty()) {
+            int amount = getAvailableManaProducers(game).size() - ability.getManaCosts().manaValue();
+            if (amount > 0) {
+                ability = ability.copy();
+                ability.getManaCostsToPay().add(new GenericManaCost(RandomUtil.nextInt(amount)));
             }
-            // check if ability kills player, if not then it's ok to play
-//            if (ability.isUsesStack()) {
-//                Game testSim = game.copy();
-//                activateAbility((ActivatedAbility) ability, testSim);
-//                StackObject testAbility = testSim.getStack().pop();
-//                testAbility.resolve(testSim);
-//                testSim.applyEffects();
-//                testSim.checkStateAndTriggered();
-//                if (!testSim.getPlayer(playerId).hasLost()) {
-//                    break;
-//                }
-//            }
-//            else {
-            break;
-//            }
```

Comment:
What was going on with this block of commented out code?

Reply:
> What was going on with this block of commented out code?

@Alex-Vasile this block of commented out code was added via [this commit](https://github.com/magefree/mage/commit/142207ce86c49c1e588ee6db0484bb624c36eb35) in 2012. This commit created the class in the first place, so the code was already commented out upon first being merged into master. The author didn't provide any additional context with the commit but it looks to be as line 99 describes: this would prevent the random player from taking an action which directly results in it losing the game.

Reply:
We'd want a way to keep trying playable abilities until either we find one that doesn't kill us, or we exhaust the list (and arbitrarily choose the ability which kills us). We also might want to do something similar within the playable options of a chosen ability, in case there is a choice of options which doesn't kill us.

The easiest way might be to remove abilities from the list until we find one that doesn't kill. Or we can maybe keep a boolean flag somewhere or each playable ability (a map, or since we are dealing with a list, we could have a list of same size). Each time we try an ability, if it kills the player, set the flag for that value to true.

Reply:
> We'd want a way to keep trying playable abilities until either we find one that doesn't kill us, or we exhaust the list (and arbitrarily choose the ability which kills us). We also might want to do something similar within the playable options of a chosen ability, in case there is a choice of options which doesn't kill us.
> 
> The easiest way might be to remove abilities from the list until we find one that doesn't kill. Or we can maybe keep a boolean flag somewhere or each playable ability (a map, or since we are dealing with a list, we could have a list of same size). Each time we try an ability, if it kills the player, set the flag for that value to true.

Yes. I'm thinking we copy the list, pop one at random saving the index, see if it kills the ai. If it doesn't then return the ability in the original list at the index. If it does, keep repeating. If we find nothing, then return one at random.

Reply:
In my opinion, a truly random agent should have equal probability of taking any legal actions, regardless of whether the action results in a loss of the game. That being said, it should definitely be restricted to taking actions that are legal. There are actions you aren't allowed to take which cause you to lose, eg you can't play a shockland untapped if you don't have 2 life to pay. On the other hand there are actions you are allowed to take which cause you to lose the game, such as playing Thoughtseize at 1 life or casting Shock on yourself. Rather than this loop, verification of what game-losing actions are legal would probably require investigating the `getPlayable()` method that is invoked.

That being said, I think the method you outline here would work for preventing this agent from taking actions that lose it the game, and I'll close this PR if y'all would prefer to make that change.

Reply:
I agree with you. But this AI seems to only be used for testing cards at random. So I think that having it play randomly for as long as possible would better fit that role.

Reply:
> In my opinion, a truly random agent should have equal probability of taking any legal actions, regardless of whether the action results in a loss of the game. That being said, it should definitely be restricted to taking actions that are legal. There are actions you aren't allowed to take which cause you to lose, eg you can't play a shockland untapped if you don't have 2 life to pay. On the other hand there are actions you are allowed to take which cause you to lose the game, such as playing Thoughtseize at 1 life or casting Shock on yourself. Rather than this loop, verification of what game-losing actions are legal would probably require investigating the `getPlayable()` method that is invoked.
> 
> 
> 
> That being said, I think the method you outline here would work for preventing this agent from taking actions that lose it the game, and I'll close this PR if y'all would prefer to make that change.

Do you want implement that in this PR yourself? Or should we open an issue to track this for later?

Reply:
> I agree with you. But this AI seems to only be used for testing cards at random. So I think that having it play randomly for as long as possible would better fit that role.

With respect to testing cards at random, imo it is a very low percentage of the time where the agent is able to take an action that directly results in losing it the game. And if there are bugs in legal routes for a player to lose themselves the game, this agent would not be able to pick up on these bugs if we instruct it to avoid these routes.

This is just my opinion and it's no more valid than yours, and either way the only test which invokes this agent currently has an `@Ignore` tag so it's not ran anyway 🤷‍♀️

Reply:
@JayDi85, do you ahve some insight in what the purpose of the random player tests are? Are they worth keeping at all?

Reply:
It's same code from `SimulatedPlayerMCTS`. Also see stack related checking in `triggerAbility` code too (search by `ability = options.get(0);`). So if you change a copy-pasted code in one place then must change code in another place too.

`RandomPlayer` doesn't used in games or other tests -- it's just a testing code for a new feature (someone improved AI code and made some additional).  I don't know is `RandomPlayer` is workable or not now. Same for `SimulatedPlayerMCTS` -- it's not a real "playable agent" -- it's just a testing code for new AI technology.

Current AI "agent" in `ComputerPlayer7` uses both AI technologies (random plays and MCTS tree with score system). So if you want to use outdated agents then must specify why you need it and what you want to do (see example with MCTS in #7075). If it's just an IntelliJ warning about "unused loop" then ignore and don't change the code.